### PR TITLE
feat: open generated project as multi root

### PR DESCRIPTION
**Pull request by slavik-lvovsky**

### CODE REVIEW

File: packages/backend/src/vscode-youi-events.ts [hunk 1]
```diff
@@ -199,17 +180,61 @@ export class VSCodeYouiEvents implements YouiEvents {
       if (selectedWorkspace === this.messages.open_in_a_new_workspace) {
         void vscode.commands.executeCommand("vscode.openFolder", targetFolderUri);
       } else if (selectedWorkspace === this.messages.add_to_workspace) {
-        const wsFoldersQuantity = _.size(vscode.workspace.workspaceFolders);
+        const wsFoldersQuantity = size(vscode.workspace.workspaceFolders);
         vscode.workspace.updateWorkspaceFolders(wsFoldersQuantity, null, {
           uri: targetFolderUri,
         });
+        if (wsFoldersQuantity === 0) {
+          const workspaceFileUri = this.createNewWorkspaceFile(targetFolderUri.fsPath);
```
Comment (alex-gilin) [inline packages/backend/src/vscode-youi-events.ts:188]: potential null reference when targetFolderPath==undefined

File: packages/backend/src/vscode-youi-events.ts [hunk 2]
```diff
@@ -199,17 +180,61 @@ export class VSCodeYouiEvents implements YouiEvents {
...
     return vscode.window.showErrorMessage(errorMmessage);
   }
 
+  private createNewWorkspaceFile(targetFolderPath: string): vscode.Uri {
+    const wsFilePath = this.getUniqWorkspaceFilePath();
+    // targetFolderPath should contain only slashes
+    // all backlsashes are replaced
+    const fileContent: string = `{
+      "folders": [{
+        "path": "${targetFolderPath.replace(/\\/g, "/")}"
```
Comment (alex-gilin) [inline packages/backend/src/vscode-youi-events.ts:204]: maybe better to resolve the path by path.resolve(...), which also replaces '\\' and expands the relative paths

File: packages/backend/src/vscode-youi-events.ts [hunk 3]
```diff
@@ -199,17 +180,61 @@ export class VSCodeYouiEvents implements YouiEvents {
...
+    const fileContent: string = `{
+      "folders": [{
+        "path": "${targetFolderPath.replace(/\\/g, "/")}" 
+      }],
+      "settings": {
+        "actions": []
+      }
+  }`;
+
+    fs.writeFileSync(wsFilePath, fileContent);
```
Comment (alex-gilin) [inline packages/backend/src/vscode-youi-events.ts:211]: may throws exception, isn't ?

File: packages/backend/src/vscode-youi-events.ts [hunk 4]
```diff
@@ -199,17 +180,61 @@ export class VSCodeYouiEvents implements YouiEvents {
...
+  private createWsFilePath(counter?: number): string {
+    const wsFileExt = this.isInBAS ? "theia-workspace" : "code-workspace";
+    const wsFileName = this.getWsFileName();
+    const counterStr = counter ? `.${counter}.` : `.`;
+    return path.join(YeomanUI["PROJECTS"], `${wsFileName}${counterStr}${wsFileExt}`);
+  }
+
+  private getWsFileName(): string {
+    const defaultName = "workspace";
+    return this.isInBAS ? get(process, "env.WORKSPACE_ID", defaultName) : get(process, "env.USERNAME", defaultName);
```
Comment (alex-gilin) [inline packages/backend/src/vscode-youi-events.ts:224]:
```suggestion
    return get(process, this.isInBAS ? "env.WORKSPACE_ID" : "env.USERNAME", defaultName);
```

File: packages/backend/test/yeomanui.spec.ts [hunk 5]
```diff
@@ -310,7 +310,8 @@ describe("yeomanui unit test", () => {
   it("_notifyGeneratorsChange", async () => {
     envUtilsMock.expects("getGeneratorsData").resolves();
     rpcMock.expects("invoke").withArgs("updateGeneratorsPrompt");
-    _.set(vscode, "workspace.workspaceFolders[0].uri.fsPath", undefined);
+    wsConfigMock.expects("get").withExactArgs("ApplicationWizard.Workspace").returns({});
+    // _.set(vscode, "workspace.workspaceFolders[0].uri.fsPath", undefined);
```
Comment (alex-gilin) [inline packages/backend/test/yeomanui.spec.ts:314]: to remove ?

File: packages/backend/src/vscode-youi-events.ts (hunk 2 above)
Comment (slavik-lvovsky) [reply packages/backend/src/vscode-youi-events.ts]: it should be \\, neither \ nor / that is why resolve does not help

File: packages/backend/src/vscode-youi-events.ts (hunk 3 above)
Comment (slavik-lvovsky) [reply packages/backend/src/vscode-youi-events.ts:210]: no,
1.  it overrides an existing file if it exists
2. It is always created in user/projects folder that is writable

File: packages/backend/src/vscode-youi-events.ts [hunk 6]
```diff
@@ -199,17 +180,61 @@ export class VSCodeYouiEvents implements YouiEvents {
...
+  }`;
+
+    fs.writeFileSync(wsFilePath, fileContent);
+    return vscode.Uri.file(wsFilePath);
+  }
+
+  private createWsFilePath(counter?: number): string {
+    const wsFileExt = this.isInBAS ? "theia-workspace" : "code-workspace";
+    const wsFileName = this.getWsFileName();
+    const counterStr = counter ? `.${counter}.` : `.`;
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply packages/backend/src/vscode-youi-events.ts:217]: no, i need .<number>. or only .

File: packages/backend/src/vscode-youi-events.ts (hunk 1 above)
Comment (slavik-lvovsky) [reply packages/backend/src/vscode-youi-events.ts:188]: no, this object always returns non empty values

File: packages/backend/test/yeomanui.spec.ts (hunk 5 above)
Comment (slavik-lvovsky) [reply packages/backend/test/yeomanui.spec.ts:314]: Done

**Review by alex-gilin — APPROVED** [review]

**Pull request merged**